### DISCUSSION TITLE
[RFC] Respect nullability type hint in response handling

### DISCUSSION THREAD

**Discussion by kuhnroyal** [cfug/dio#1335] [Ideas]
Currently we execute requests with `dio.get<T>()` and the response data is always of type `T?`.

In the future we could respect the type hint and return response data of type `T`.

1. `dio.get<T>()` => `T`
2. `dio.get<T?>()` => `T?` 

For case 1 we would have to add error handling for the special case of the response unexpectedly being empty.

This would be a breaking change and thus would require a major version bump.

This was first raised in https://github.com/flutterchina/dio/pull/1292

**Comment by kuhnroyal** [comment]
@wendux @jgoyvaerts What do you guys think?

**Comment by AlexV525** [comment]
Even we strict the generic type to T, the T itself extends `Object?` by default. For somehow it's not helping the nullability, but can be a improvement of static analysis.

**Reply by kuhnroyal** [reply]
Yea that's true.

**Reply by kuhnroyal** [reply]
Actually that is not really true. We can provide the correct response type. This is a breaking change but we could target this for a 6.0 release: https://github.com/cfug/dio/pull/1755
